docs/src/adoc/index.adoc

```diff
@@ -659,24 +919,159 @@ Out of the box, Jdbi supports mapping database columns to the following types:
   `OffsetDateTime`, `ZonedDateTime`, and `ZoneId`
 * java.util: `UUID`
 * `java.util.Collection` and Java arrays (for array columns). Some
-  additional setup may be required depending on the type of array element.
+  additional setup may be required depending on the type of array element--see
+  <<SQL Arrays>>.
 
-TODO:
+===== ColumnMapperFactory
 
-* Registering column mappers allows Jdbi to automatically convert database
-  result columns to the mapped type.
-* Registering an inline column mapper by mapped type
-* Registering a concretely typed column mapper
-* Last registered mapper for a given type wins. This means you can override the
-  out-of-the-box mapper for any type with your own mapper
+A link:{jdbidocs}/core/mapper/ColumnMapperFactory.html[ColumnMapperFactory^] can
+produce column mappers for arbitrary types.
 
-===== ColumnMapperFactory
+Implementing a factory might be preferable to a regular column mapper if:
 
-TODO:
+* The mapper class is generic, and could apply to multiple mapped types.
+* The type being mapped is generic, and/or the mapper could be composed
+  from other registered mappers.
+* You want to bundle multiple mappers into a single class.
+
+Let's create a mapper factory for `Optional<T>` as an example. The factory
+should produce a `ColumnMapper<Optional<T>>` for any `T`, provided a column
+mapper is registered for `T`.
 
-* For mapping arbitrary value types, e.g. Money
-* How to implement a column mapper factory
-* Use GenericTypes helper to help with generics.
+Let's take this one step at a time:
+
+[source,java]
+----
+public class OptionalColumnMapperFactory implements ColumnMapperFactory {
+  public Optional<ColumnMapper<?>> build(Type type, ConfigRegistry config) {
+    ...
+  }
+}
+----
+
+The `build` method accepts a mapped type, and a config registry. It may return
+`Optional.of(someMapper)` if it knows how to map that type, or
+`Optional.empty()` otherwise.
+
+First, we check whether the mapped type is an `Optional`:
+
+[source,java]
+----
+if (!Optional.class.equals(GenericTypes.getErasedType(type))) {
+  return Optional.empty();
+}
+----
+
+The link:{jdbidocs}/core/generic/GenericTypes.html[GenericTypes^] utility class
+provides several methods for helping to interpret Java generic types.
+
+The method `GenericTypes.getErasedType` accepts a `java.lang.reflect.Type` and
+returns the raw `java.lang.Class` for that type. So if `type` were the generic
+type `Optional<String>`, it would return `Optional.class`.
+
+Next, we extract the `T` generic parameter from the mapped type:
+
+[source,java]
+----
+Type embeddedType = GenericTypes.resolveType(Optional.class.getTypeParameters()[0], type);
+----
+
+The `GenericTypes.resolveType` method accepts a generic type parameter to
+resolve, and a context type in which to resolve it.
+
+The expression `Optional.class.getTypeParameters()[0]` gives the type variable
```

Comment:
This is great explanation, but it feels repetitive from the same text around line 723.  Maybe we should just have a section on generic type handling and refer it from the two locations?

Reply:
Will do